### PR TITLE
Emit build operations for early executed artifact transforms

### PR DESCRIPTION
**Pull request by lptr**

### CODE REVIEW

**Comment by lptr** [comment]
Branch build: https://builds.gradle.org/viewQueued.html?itemId=14018231

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/transform/TransformInfoExecutor.java [hunk 1]
```diff
@@ -18,13 +18,38 @@
...
+            final TransformInfo transform = (TransformInfo) work;
+            buildOperationExecutor.run(new RunnableBuildOperation() {
+                @Override
+                public void run(BuildOperationContext context) {
+                    transform.execute();
+                }
+
+                @Override
+                public BuildOperationDescriptor.Builder description() {
+                    String displayName = "Transform " + transform.toString();
```
Comment (wolfs) [inline subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/transform/TransformInfoExecutor.java]: For combined transforms the arrow seems to be the wrong way: 
```
> Transform FindRelocatedClasses -> ShadeClassesTransform -> MinifyTransform on groovy-all.jar (org.codehaus.groovy:groovy-all:2.4.12)
```

The data flow is in the opposite direction of the arrows. Do you think the other arrow direction would be better?

**Comment by lptr** [comment]
Can you please reproduce that error with `--stacktrace`?

**Comment by wolfs** [comment]
@lptr `--stacktrace` is already added as seen in the screen cast. The stacktrace still doesn't get printed out...
Does that mean you cannot reproduce it? Did you build a local distribution and run `:toolingApi:toolingApiShadedJar` with it?

**Comment by oehme** [comment]
@lptr This seems to be the last issue for RC2, what's the current status?

**Comment by lptr** [comment]
There's some work to be done still. I intend to merge it today by late EU afternoon.

**Comment by lptr** [comment]
Branch build: https://builds.gradle.org/viewQueued.html?itemId=14115195

**Comment by wolfs** [comment]
@lptr Seems like you merged `master` into this branch instead of `release`.

**Comment by lptr** [comment]
Let's try again... https://builds.gradle.org/viewQueued.html?itemId=14115596

**Comment by lptr** [comment]
Hm, I can reproduce the:

```text
* What went wrong:
Received complete event for an unknown operation (id: 14257)
```

...problem when applying the changes to `release`, but not when doing the same on `master`...

**Comment by lptr** [comment]
OK, it's not about where I apply the changes: I was previously building an older version of Gradle that was not using artifact transforms...

**Comment by lptr** [comment]
Looks like the problem is caused by classifying transform operations as `BuildOperationCategory.TASK`. Without that everything works fine.

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/transform/TransformInfo.java [hunk 2]
```diff
@@ -200,11 +149,16 @@ protected ResolvedTransformInputs resolveInputs() {
...
-                ResolvedArtifactResult artifact = Iterables.getOnlyElement(visitor.getArtifacts());
-                return new ResolvedTransformInputs(ImmutableList.of(artifact.getFile()));
+                this.failure = failure;
+                this.result = Collections.emptyList();
+                return;
             }
+            ResolvedArtifactResult artifact = Iterables.getOnlyElement(visitor.getArtifacts());
+
+            TransformArtifactOperation operation = new TransformArtifactOperation(artifact.getId(), artifact.getFile(), artifactTransformer);
+            buildOperationExecutor.run(operation);
```
Comment (wolfs) [inline subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/transform/TransformInfo.java]: Should we do the check if the `transform` has a cached result here, too? So we only would emit build operations when the artifact is actually transformed?

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/transform/TransformArtifactOperation.java [hunk 3]
```diff
@@ -29,31 +30,35 @@
...
-    TransformArtifactOperation(ResolvableArtifact artifact, ArtifactTransformer transform) {
-        this.artifact = artifact;
+    TransformArtifactOperation(ComponentArtifactIdentifier artifactId,  File file, ArtifactTransformer transform) {
+        this.artifactId = artifactId;
+        this.file = file;
         this.transform = transform;
     }
 
     @Override
     public void run(@Nullable BuildOperationContext context) {
```
Comment (wolfs) [inline subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/transform/TransformArtifactOperation.java]: `context` should be `Nullable` in `TransformFileOperation`, too.

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/transform/TransformArtifactOperation.java [hunk 4]
```diff
@@ -29,31 +30,35 @@
...
         } catch (Throwable t) {
             failure = t;
         }
     }
 
     @Override
     public BuildOperationDescriptor.Builder description() {
-        return BuildOperationDescriptor.displayName("Apply " + transform.getDisplayName() + " to " + artifact);
+        return BuildOperationDescriptor.displayName("Transform " + artifactId.getDisplayName() + " with " + transform.getDisplayName())
+            .progressDisplayName("Transform " + transform.getDisplayName())
```
Comment (wolfs) [inline subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/transform/TransformArtifactOperation.java]: I think we should add the `file` to the progress display name, too. If the file is not there, we only see something like `Transform MinifyTransform` which is not very helpful IMO.

**Comment by wolfs** [comment]
I also guess that `BuildScanPluginFixture` needs to move to the `testFixtures` source set.

**Comment by lptr** [comment]
Aaand again: https://builds.gradle.org/viewLog.html?buildId=14119200

**Comment by lptr** [comment]
I mean: https://builds.gradle.org/viewQueued.html?itemId=14119488

**Comment by lptr** [comment]
Next: https://builds.gradle.org/viewQueued.html?itemId=14120328

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/transform/TransformInfo.java (hunk 2 above)
Comment (lptr) [reply subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/transform/TransformInfo.java]: I don't think we should. We'll revamp the whole caching architecture anyway.

@oehme do you know how much effect this `if` has?

https://github.com/gradle/gradle/blob/3ba5478382886306fc5bd62c956a28701011159e/subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/transform/TransformingAsyncArtifactListener.java#L50-L54

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/transform/TransformArtifactOperation.java (hunk 3 above)
Comment (lptr) [reply subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/transform/TransformArtifactOperation.java]: Fixed.

File: subprojects/dependency-management/src/integTest/groovy/org/gradle/api/artifacts/transform/ArtifactTransformBuildScanIntegrationTest.groovy [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+            include 'lib'
+            include 'util'
+        """
+        fixture.logConfig = true
+        fixture.publishDummyBuildScanPluginNow()
+    }
+
+    def "transform works with build scan"() {
+        given:
+        buildFile << """
```
Comment (wolfs) [inline subprojects/dependency-management/src/integTest/groovy/org/gradle/api/artifacts/transform/ArtifactTransformBuildScanIntegrationTest.groovy]: Can we now re-use code from the other artifact transform tasks so we don't need to add this large code block again?

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/transform/TransformArtifactOperation.java (hunk 4 above)
Comment (lptr) [reply subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/transform/TransformArtifactOperation.java]: Added. I think the most important part is that we see _something_ happening. We can figure out the best way to show progress based on real-world experience.

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/transform/TransformInfo.java [hunk 6]
```diff
@@ -200,11 +150,16 @@ protected ResolvedTransformInputs resolveInputs() {
...
-            } else {
-                ResolvedArtifactResult artifact = Iterables.getOnlyElement(visitor.getArtifacts());
-                return new ResolvedTransformInputs(ImmutableList.of(artifact.getFile()));
+                this.failure = failure;
+                this.result = Collections.emptyList();
+                return;
             }
+            ResolvedArtifactResult artifact = Iterables.getOnlyElement(visitor.getArtifacts());
+
+            TransformArtifactOperation operation = new TransformArtifactOperation(artifact.getId(), artifact.getFile(), artifactTransformer, BuildOperationCategory.TASK);
```
Comment (wolfs) [inline subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/transform/TransformInfo.java]: What happens if we use `UNCATEGORIZED` here?

**Comment by lptr** [comment]
One more: https://builds.gradle.org/viewQueued.html?itemId=14120922

**Comment by lptr** [comment]
@wolfs please take another look.

**Comment by lptr** [comment]
Rerun: https://builds.gradle.org/viewQueued.html?itemId=14125625

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/transform/TransformInfo.java (hunk 6 above)
Comment (lptr) [reply subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/transform/TransformInfo.java]: It disappears from the verbose console.

**Comment by lptr** [comment]
I'm going to run another quick feedback build for the last change ("display only file name for file transforms"), and then merge this. [The only performance test failing is flaky](https://builds.gradle.org/viewLog.html?buildId=14125711&buildTypeId=Gradle_Check_IndividualPerformanceScenarioWorkersLinux), and everything else is fine.

**Pull request merged**